shared/src/balancer/logexpmath.rs

```diff
@@ -0,0 +1,520 @@
+//! Module emulating the pow function as found in the Balancer contract code.
+// The original contract code can be found at:
+// https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/solidity-utils/contracts/math/LogExpMath.sol
+
+#![allow(dead_code)]
+
+use ethcontract::{I256, U256};
+use lazy_static::lazy_static;
+use std::convert::TryFrom;
+
+lazy_static! {
+    static ref ONE_18: I256 = I256::from(10_u128.pow(18));
+    static ref ONE_20: I256 = I256::from(10_u128.pow(20));
+    static ref ONE_36: I256 = I256::from(10_u128.pow(36));
+    static ref UNSIGNED_ONE_18: U256 = U256::try_from(*ONE_18).unwrap();
+    static ref MAX_NATURAL_EXPONENT: I256 = ONE_18.checked_mul(I256::from(130_i128)).unwrap();
+    static ref MIN_NATURAL_EXPONENT: I256 = ONE_18.checked_mul(I256::from(-41_i128)).unwrap();
+    static ref LN_36_LOWER_BOUND: I256 = ONE_18.checked_sub(I256::from(10_u128.pow(17))).unwrap();
+    static ref LN_36_UPPER_BOUND: I256 = ONE_18.checked_add(I256::from(10_u128.pow(17))).unwrap();
+    static ref MILD_EXPONENT_BOUND: I256 =
+        I256::from(2_u32).pow(254_u32).checked_div(*ONE_20).unwrap();
+}
+fn constant_x_20(i: u32) -> I256 {
+    match i {
+        2 => 3_200_000_000_000_000_000_000_i128,
+        3 => 1_600_000_000_000_000_000_000_i128,
+        4 => 800_000_000_000_000_000_000_i128,
+        5 => 400_000_000_000_000_000_000_i128,
+        6 => 200_000_000_000_000_000_000_i128,
+        7 => 100_000_000_000_000_000_000_i128,
+        8 => 50_000_000_000_000_000_000_i128,
+        9 => 25_000_000_000_000_000_000_i128,
+        10 => 12_500_000_000_000_000_000_i128,
+        11 => 6_250_000_000_000_000_000_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_x_18(i: u32) -> I256 {
+    match i {
+        0 => 128_000_000_000_000_000_000_i128,
+        1 => 64_000_000_000_000_000_000_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_a_20(i: u32) -> I256 {
+    match i {
+        2 => 7_896_296_018_268_069_516_100_000_000_000_000_i128,
+        3 => 888_611_052_050_787_263_676_000_000_i128,
+        4 => 298_095_798_704_172_827_474_000_i128,
+        5 => 5_459_815_003_314_423_907_810_i128,
+        6 => 738_905_609_893_065_022_723_i128,
+        7 => 271_828_182_845_904_523_536_i128,
+        8 => 164_872_127_070_012_814_685_i128,
+        9 => 128_402_541_668_774_148_407_i128,
+        10 => 113_314_845_306_682_631_683_i128,
+        11 => 106_449_445_891_785_942_956_i128,
+        _ => panic!("Constant not provided"),
+    }
+    .into()
+}
+fn constant_a_18(i: u32) -> I256 {
+    match i {
+        0 => {
+            I256::from_dec_str("38877084059945950922200000000000000000000000000000000000").unwrap()
+        }
+        1 => 6_235_149_080_811_616_882_910_000_000_i128.into(),
+        _ => panic!("Constant not provided"),
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Errors {
```

Comment:
Probably would be nice to use https://docs.rs/thiserror/1.0.25/thiserror/ to derive Error for this type. But depends on how the function that returns this is used so maybe it's not needed if we never bubble them up or turn them into anyhow::Error.

Reply:
Eventually all these errors will disappear into an `Option`. This enum is going to be extended to include all Balancer revert errors that we need, so that I can work with other contract functions reverting using `?`. So unless things change I'd keep it like this without an explicit error message provided by `thiserror`.

Reply:
> so that I can work with other contract functions reverting using ?.

In this case, we probably want to use `thiserror` as `anyhow` requires errors to implement `Error` (or at least `Display`) if I'm not mistaken. Also, in general, I think its nice that error types implement `std::error::Error`.

Reply:
This error struct will be extended to contain other Balancer errors, so `anyhow` should not be required. On the other hand, having a backtrace of the error would be nice, since the same error will be thrown from different functions in the code as in the balancer contract. I'll make the changes when I'll introduce more errors!

Reply:
> , so anyhow should not be required

But if the calling code wants to handle the error with `?` and `anyhow` then it would be.

> I'll make the changes when I'll introduce more errors!

Sounds good to me!